### PR TITLE
fs/inode: improve the performance of get file pointer

### PR DESCRIPTION
**Pull request by anchao**
## Summary

1. fs/inode: improve the performance of get file pointer

Remove file locks and enter critical sections to protect file lists

Signed-off-by: chao an <anchao@xiaomi.com>

2. fs/inode: minor change to save the code size

```
Before:
   text    data     bss     dec     hex filename
 360641    7889    4064  372594   5af72 nuttx

After:
   text    data     bss     dec     hex filename
 360313    7889    4064  372266   5ae2a nuttx
```

Signed-off-by: chao an <anchao@xiaomi.com>

## Impact

N/A

## Testing

ci-check

### CODE REVIEW

File: fs/inode/fs_files.c [hunk 1]
```diff
@@ -105,9 +106,16 @@ static int files_extend(FAR struct filelist *list, size_t row)
     }
   while (++i < row);
 
+  flags = enter_critical_section();
```
Comment (xiaoxiang781216) [inline fs/inode/fs_files.c:109]: the sequence to hold the critical section and fl_lock is difference from fs_getfilep, the deadlock will happen on SMP machine.

Comment (anchao) [reply fs/inode/fs_files.c:133]: Done, I will remove file locks and use critical sections to protect file lists

File: fs/inode/fs_files.c [hunk 2]
```diff
@@ -51,59 +51,118 @@
...
+ * Name: files_fget_index
+ ****************************************************************************/
+
+static FAR struct file *files_fget_index(FAR struct filelist *list,
+                                         int l1, int l2)
+{
+  FAR struct file *filep;
+  irqstate_t flags;
+
+  flags = enter_critical_section();
```
Comment (raiden00pl) [inline fs/inode/fs_files.c:64]: shouldn't we use spin_lock_irqsave here and below ?

File: fs/inode/fs_files.c [hunk 3]
```diff
@@ -278,10 +316,28 @@ void files_releaselist(FAR struct filelist *list)
...
+ * Returned Value:
+ *   Zero (OK) is returned on success; a negated errno value is returned on
+ *   any failure.
+ *
+ ****************************************************************************/
 
-  nxmutex_destroy(&list->fl_lock);
+int files_get_filecount(FAR struct filelist *list)
+{
+  return list->fl_rows * CONFIG_NFILE_DESCRIPTORS_PER_BLOCK;
```
Comment (raiden00pl) [inline fs/inode/fs_files.c:340]: what about protection `list->fl_rows` from race conditions? I know that `list->fl_rows` will rarely be changed, but won't this make any bugs even rarer and harder to detect? or is there no negative consequences from this ?

**Review by raiden00pl — COMMENTED** [review]
An opinion from someone familiar with SMP would be useful here.
Maybe we can design a test application that will be able to detect possible problems in a real HW?

File: fs/inode/fs_files.c (hunk 2 above)
Comment (anchao) [reply fs/inode/fs_files.c:64]: `critical_section` has been protected by spin lock in the case of SMP

File: fs/inode/fs_files.c (hunk 3 above)
Comment (anchao) [reply fs/inode/fs_files.c:340]: `fl_rows` will not shrink dynamically, there will be no side effect if it continues to increase.

**Comment by raiden00pl** [comment]
BTW this PR quite drastically reduces the maximum control loop duration for FOC (tested with stm32f405rb which I currently have on the bench). Before this PR: 5502 cycles, with this PR: 4869 cycles

File: fs/inode/fs_files.c (hunk 2 above)
Comment (xiaoxiang781216) [reply fs/inode/fs_files.c:64]: I think @raiden00pl suggest that spinlock can be faster in SMP case, so it's better to change mutex to spinlock_t

File: fs/inode/fs_files.c [hunk 4]
```diff
@@ -51,14 +51,47 @@
...
+  leave_critical_section(flags);
+
+  return filep;
+}
+
+/****************************************************************************
+ * Name: files_fget
+ ****************************************************************************/
+
+static FAR struct file *files_fget(FAR struct filelist *list, int fd)
```
Comment (xiaoxiang781216) [inline fs/inode/fs_files.c:77]: files_fget->files_get

File: fs/inode/fs_files.c [hunk 5]
```diff
@@ -51,14 +51,47 @@
  * Private Functions
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: files_fget_index
+ ****************************************************************************/
+
+static FAR struct file *files_fget_index(FAR struct filelist *list,
```
Comment (xiaoxiang781216) [inline fs/inode/fs_files.c:58]: files_fget_index->files_get_by_index

File: include/nuttx/fs/fs.h [hunk 6]
```diff
@@ -850,6 +850,19 @@ void files_initlist(FAR struct filelist *list);
...
+ *
+ * Description:
+ *   Get file count from file list
+ *
+ * Returned Value:
+ *   file count of file list
+ *
+ ****************************************************************************/
+
+int files_get_filecount(FAR struct filelist *list);
```
Comment (xiaoxiang781216) [inline include/nuttx/fs/fs.h:864]: files_countlist, files_get_count or files_get_number

File: fs/inode/fs_files.c [hunk 7]
```diff
@@ -51,59 +51,118 @@
...
-  /* Note: If assertion occurs, the fl_rows has a overflow.
-   * And there may be file descriptors leak in system.
+  /* To avoid race condition, if the file list is updated by other threads
+   * and list rows is greater or equal than temp list,
+   * release the obsolete buffers
    */
 
-  DEBUGASSERT(list->fl_rows == row);
-  return 0;
+  if (orig_rows != list->fl_rows && list->fl_rows >= row)
```
Comment (xiaoxiang781216) [inline fs/inode/fs_files.c:140]:
```suggestion
  if (orig_rows != list->fl_rows)
```
and change list->fl_rows >= row to DEBUGASSERT in if block.

File: fs/inode/fs_files.c (hunk 2 above)
Comment (anchao) [reply fs/inode/fs_files.c]: Done

File: fs/inode/fs_files.c (hunk 4 above)
Comment (anchao) [reply fs/inode/fs_files.c:78]: I do not like files_get , files_fget align with linux kernel

File: fs/inode/fs_files.c [hunk 8]
```diff
@@ -51,59 +52,118 @@
...
+}
+
+/****************************************************************************
+ * Name: files_fget
+ ****************************************************************************/
+
+static FAR struct file *files_fget(FAR struct filelist *list, int fd)
+{
+  return files_fget_by_index(list, fd / CONFIG_NFILE_DESCRIPTORS_PER_BLOCK,
+                          fd % CONFIG_NFILE_DESCRIPTORS_PER_BLOCK);
```
Comment (xiaoxiang781216) [inline fs/inode/fs_files.c:81]: align

File: fs/inode/fs_files.c [hunk 9]
```diff
@@ -51,59 +52,118 @@
...
+ * Name: files_fget_by_index
+ ****************************************************************************/
+
+static FAR struct file *files_fget_by_index(FAR struct filelist *list,
+                                         int l1, int l2)
+{
+  FAR struct file *filep;
+  irqstate_t flags;
+
+  flags = spin_lock_irqsave(NULL);
```
Comment (xiaoxiang781216) [inline fs/inode/fs_files.c:65]: how about change fl_lock from mutex to spinlock to improve the efficient in SMP?

Comment (TaiJuWu) [reply fs/inode/fs_files.c:65]: Can add a lock only for file?

File: fs/inode/fs_files.c (hunk 8 above)
Comment (anchao) [reply fs/inode/fs_files.c:81]: Done

File: fs/inode/fs_files.c (hunk 9 above)
Comment (anchao) [reply fs/inode/fs_files.c]: Done

File: fs/inode/fs_files.c [hunk 10]
```diff
@@ -51,14 +51,47 @@
  * Private Functions
  ****************************************************************************/
 
+/****************************************************************************
+ * Name: files_fget_by_index
+ ****************************************************************************/
+
+static FAR struct file *files_fget_by_index(FAR struct filelist *list,
+                                         int l1, int l2)
```
Comment (xiaoxiang781216) [inline fs/inode/fs_files.c:59]: align with `(`

File: fs/inode/fs_files.c [hunk 11]
```diff
@@ -756,28 +714,18 @@ int nx_close_from_tcb(FAR struct tcb_s *tcb, int fd)
...
-    {
-      return ret;
-    }
-
   /* If the file was properly opened, there should be an inode assigned */
 
   if (fd < 0 || fd >= list->fl_rows * CONFIG_NFILE_DESCRIPTORS_PER_BLOCK ||
-      !list->fl_files[fd / CONFIG_NFILE_DESCRIPTORS_PER_BLOCK]
-                     [fd % CONFIG_NFILE_DESCRIPTORS_PER_BLOCK].f_inode)
+      files_fget(list, fd)->f_inode == NULL)
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply fs/inode/fs_files.c:745]: it's minor change, why not done in this patch?

[the remark this answers is not in the document]
Comment (anchao) [reply fs/inode/fs_files.c]: Done

File: fs/inode/fs_files.c (hunk 10 above)
Comment (anchao) [reply fs/inode/fs_files.c]: Done

File: include/nuttx/spinlock.h [hunk 12]
```diff
@@ -32,12 +32,12 @@
 
 #include <nuttx/irq.h>
 
-#ifdef CONFIG_RW_SPINLOCK
-#include <stdatomic.h>
+#if defined(CONFIG_RW_SPINLOCK) && !defined(__cplusplus)
```
Comment (xiaoxiang781216) [inline include/nuttx/spinlock.h:35]:
```
#ifdef CONFIG_RW_SPINLOCK
#  ifdef __cplusplus
#    include <atomic.h>
#  else
#    include <stdatomic.h>
#  endif
```

File: include/nuttx/spinlock.h [hunk 13]
```diff
@@ -500,7 +509,7 @@ void spin_unlock_irqrestore_wo_note(FAR spinlock_t *lock, irqstate_t flags);
 #  define spin_unlock_irqrestore_wo_note(l, f) up_irq_restore(f)
 #endif
 
-#ifdef CONFIG_RW_SPINLOCK
+#if defined(CONFIG_RW_SPINLOCK) && !defined(__cplusplus)
```
Comment (xiaoxiang781216) [inline include/nuttx/spinlock.h:512]: revert the change

File: include/nuttx/spinlock.h (hunk 12 above)
Comment (anchao) [reply include/nuttx/spinlock.h:35]: > In file included from /home/archer/code/nuttx/n1/nuttx/include/nuttx/fs/fs.h:40,
>                  from /home/archer/code/nuttx/n1/nuttx/include/nuttx/lib/lib.h:30,
>                  from libcxxmini/libxx_delete.cxx:26:
> /home/archer/code/nuttx/n1/nuttx/include/nuttx/spinlock.h:37:14: fatal error: atomic.h: No such file or directory
>    37 | #    include <atomic.h>
>       |              ^~~~~~~~~~
> compilation terminated.
> make[1]: *** [Makefile:69: libxx_delete.o] Error 1
> make: *** [tools/LibTargets.mk:216: libs/libxx/libxx.a] Error 2
> make: *** Waiting for unfinished jobs....
>

Comment (anchao) [reply include/nuttx/spinlock.h:35]: What GCC version are you using?

Comment (anchao) [reply include/nuttx/spinlock.h:35]:
```
$ arm-none-eabi-gcc -v
Using built-in specs.
COLLECT_GCC=arm-none-eabi-gcc
COLLECT_LTO_WRAPPER=/home/archer/code/k03/k1/prebuilts/gcc/linux/arm/bin/../libexec/gcc/arm-none-eabi/11.3.1/lto-wrapper
Target: arm-none-eabi
Configured with: /data/jenkins/workspace/GNU-toolchain/arm-11/src/gcc/configure --target=arm-none-eabi --prefix=/data/jenkins/workspace/GNU-toolchain/arm-11/build-arm-none-eabi/install --with-gmp=/data/jenkins/workspace/GNU-toolchain/arm-11/build-arm-none-eabi/host-tools --with-mpfr=/data/jenkins/workspace/GNU-toolchain/arm-11/build-arm-none-eabi/host-tools --with-mpc=/data/jenkins/workspace/GNU-toolchain/arm-11/build-arm-none-eabi/host-tools --with-isl=/data/jenkins/workspace/GNU-toolchain/arm-11/build-arm-none-eabi/host-tools --disable-shared --disable-nls --disable-threads --disable-tls --enable-checking=release --enable-languages=c,c++,fortran --with-newlib --with-gnu-as --with-gnu-ld --with-sysroot=/data/jenkins/workspace/GNU-toolchain/arm-11/build-arm-none-eabi/install/arm-none-eabi --with-multilib-list=aprofile,rmprofile --with-pkgversion='Arm GNU Toolchain 11.3.Rel1' --with-bugurl=https://bugs.linaro.org/
Thread model: single
Supported LTO compression algorithms: zlib
gcc version 11.3.1 20220712 (Arm GNU Toolchain 11.3.Rel1) 

```

**Review by xiaoxiang781216 — APPROVED** [review]
